pythonPackages.pex: 1.2.7 -> 1.4.5

mgeisler opened the pull request:
###### Motivation for this change

This packages a new version of pex and moves it to a top-level tool, similarly to how Pipenv is packaged.

I'm not sure what the impact is of the move from `pkgs/top-level/python-packages.nix` to `pkgs/top-level/all-packages.nix`, but since pex is more of a command line tool than a Python library, the change makes sense to me :-) Please advice if this is bad for some reason or if there is followup work to be done!

The packaging was tested locally on a MacOS machine and a Ubuntu VM. The packaging was also executed on a Hydra installation with the sandbox enabled.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [x] Ubuntu
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

mgeisler commented:
> undefined variable 'pex' at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/eval-2-shlevy.ewr1.nix.ci/pkgs/development/tools/build-managers/pants/default.nix:28:61

Okay... I guess this is a hint that moving pex _does_ have an impact :-)

dotlambda commented on pkgs/development/tools/pex/default.nix:
```diff
@@ -0,0 +1,29 @@
+{ stdenv, python3Packages }:
+with python3Packages; buildPythonApplication rec {
+    name = "${pname}-${version}";
```
Drop this line

mgeisler replied:
Okay, thanks -- is it computed automatically from `pname` and `version`?

dotlambda commented on pkgs/development/tools/pex/default.nix:
```diff
@@ -0,0 +1,29 @@
+    patches = [ ./set-source-date-epoch.patch ];
+
+    # A few more dependencies I don't want to handle right now...
+    doCheck = false;
```
Did you check whether these dependencies have been packaged in the meantime?

mgeisler replied:
No, I just copied the existing definition. Removing the line seems to work fine -- though no tests are run.

Please let me know how you like to see changes made in response to review comments: as follow-up commits, or by me immediately fixing the broken commit and force pushing?

mgeisler commented on pkgs/top-level/all-packages.nix:
```diff
@@ -7620,6 +7620,7 @@ with pkgs;
 
   pew = callPackage ../development/tools/pew {};
   pipenv = callPackage ../development/tools/pipenv {};
+  pex = callPackage ../development/tools/pex {};
```
Thanks, I've inserted the package before `pew` now.

mgeisler commented:
> Do you know if it is used as an executable or as a library in pantsbuild? 

@Mic92, I looked it up and Pants do use PEX as a Python library.

So I guess that means that I should not be moving from `buildPythonPackage` to `buildPythonApplication` like I originally thought?

Mic92 commented:
Yes otherwise pantsbuild might break because no dependencies from pex to pantsbuild are propagated: https://github.com/NixOS/nixpkgs/blob/master/pkgs/top-level/python-packages.nix#L66

FRidh commented:
I don't really see the value in having a `buildPythonApplication` variant. How does it deal with multiple interpreter versions?

mgeisler commented:
> I don't really see the value in having a `buildPythonApplication` variant.

Oh, I don't know if there's any value in it either :-) I just saw that Pipenv was packaged that way and since PEX is also a command line too, I thought that using `buildPythonApplication` would be a little  more correct somehow.

I'll try updating it back to `buildPythonPackage` and see if that works better. Thanks for the tips!

mgeisler committed:
pex: move definition to python-modules/pex

mgeisler committed:
pex: 1.2.7 -> 1.4.5

The install_requires have changed from version 1.2.7 to 1.4.5 and now
also include the wheel package. So this and setuptools have been added
to propagatedBuildInputs.

mgeisler committed:
pex: handle sources from /nix/store

The files in /nix/store are all back-dated to 1970, and this breaks
pex when it tries to generate wheels: wheels are really zip files, and
the zip file format does not support dates before 1980.

This is normally handled by setting SOURCE_DATE_EPOCH when building
the Nix package. However, pex will invoke further Python interpreters
at runtime in calls like this:

  $ python setup.py bdist_wheel

We thus need to ensure that it passes the SOURCE_DATE_EPOCH variable
onto those interpreters. The date of the wheels will be fixed at 1980.

mgeisler committed:
pex: fix “Warning, using a UrllibContext which is known to be flaky”

Without requests, pex will emit a number of warnings to the terminal
when building pex files:

  % ./result/bin/pex -v httpie -o http
  pex: Warning, using a UrllibContext which is known to be flaky.
  pex: Please build pex with the requests module for more reliable downloads.
  pex: Warning, using a UrllibContext which is known to be flaky.
  pex: Please build pex with the requests module for more reliable downloads.

This fixes these warnings.

mgeisler commented on pkgs/development/python-modules/pex/default.nix:
```diff
@@ -0,0 +1,28 @@
+  propagatedBuildInputs = [ setuptools wheel requests ];
+
+  prePatch = ''
+    substituteInPlace setup.py --replace 'SETUPTOOLS_REQUIREMENT,' '"setuptools",'
```
I'm a little worried about this line -- which was part of the original derivation. The [`setup.py` file from pex][1] includes a `pex/version.py` file, which says:
```python
# NB: If we upgrade to setuptools>=34 pex's bootstrap code in `PEXBuilder` will need an update to
# include the `packaging` package in the `.bootstrap/` code since we use
# `packaging.specifiers.SpecifierSet` - indirectly - through `pkg_resources.Requirement.specifier`.
SETUPTOOLS_REQUIREMENT = 'setuptools>=20.3,<34.0'

WHEEL_REQUIREMENT = 'wheel>=0.26.0,<0.32.0'
```

The `SETUPTOOLS_REQUIREMENT` and `WHEEL_REQUIREMENT` variables are then used in `setup.py`. When we change the requirement to just `"setuptools"`, we get whatever version is in Nix -- version 40.0.0 at the moment. This sounds dangerous in light of the above comment.

Is there a normal approach in Nix to handle situations like this? I imagine one could package setuptools version 33 and have pex use that specific version?

[1]: https://github.com/pantsbuild/pex/blob/v1.4.5/setup.py

Mic92 replied:
With python we are more bound to having only single versions in nixpkgs because of dependency propagation. However it seems that master also support setuptools newer then 34: https://github.com/pantsbuild/pex/blob/f125317bb6001ccbb56f68168bc45f947d274405/pex/version.py#L8

mgeisler replied:
Ah, cool! The new `version.py` file says:
```
# Versions 34.0.0 through 35.0.2 (last pre-36.0.0) de-vendored dependencies which
# causes problems for pex code so we exclude that range.
SETUPTOOLS_REQUIREMENT = 'setuptools>=20.3,<41,!=34.*,!=35.*'
```
which means that using version 40.0.0 should be fine. Thanks for looking this up!

Mic92 commented:
@GrahamcOfBorg build pythonPackages.pex python3Packages.pex

GrahamcOfBorg commented:
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45497&attempt_id=c8692ce6-0e73-442d-bbb6-6e52ebbb1fad)

Attempted: pythonPackages.pex, python3Packages.pex

<details><summary>Partial log (click to expand)</summary><p>

```
warning: no previously-included files matching '*~' found under directory '*'
writing manifest file 'pex.egg-info/SOURCES.txt'
running build_ext

----------------------------------------------------------------------
Ran 0 tests in 0.000s

OK
/nix/store/izc6s44r6y1kpq5l1fs01f98s6b0x7db-python2.7-pex-1.4.5
/nix/store/cn7ismrgfib91vbpkqqx5y33l6m2h514-python3.6-pex-1.4.5
```
</p></details>

GrahamcOfBorg commented:
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45497&attempt_id=cd6e676f-6795-4e90-8688-45fffc4cfefe)

Attempted: pythonPackages.pex, python3Packages.pex

<details><summary>Partial log (click to expand)</summary><p>

```
Reading https://pypi.org/simple/pytest/
Download error on https://pypi.org/simple/pytest/: [Errno -2] Name or service not known -- Some packages may not be found!
Couldn't find index page for 'pytest' (maybe misspelled?)
Scanning index of all packages (this may take a while)
Reading https://pypi.org/simple/
Download error on https://pypi.org/simple/: [Errno -2] Name or service not known -- Some packages may not be found!
No local packages or working download links found for pytest
error: Could not find suitable distribution for Requirement.parse('pytest')
builder for '/nix/store/7j76m6x2l1r6942adavm7c6ry73q4fil-python3.6-pex-1.4.5.drv' failed with exit code 1
error: build of '/nix/store/7j76m6x2l1r6942adavm7c6ry73q4fil-python3.6-pex-1.4.5.drv', '/nix/store/kbvirbswfdhjjffmhx8dzmwf65cmwk8z-python2.7-pex-1.4.5.drv' failed
```
</p></details>

GrahamcOfBorg commented:
Failure on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45497&attempt_id=cbc52c84-09f9-447b-85b7-f1ae82b3a5bc)

Attempted: pythonPackages.pex, python3Packages.pex

<details><summary>Partial log (click to expand)</summary><p>

```
Reading https://pypi.org/simple/pytest/
Download error on https://pypi.org/simple/pytest/: [Errno -2] Name or service not known -- Some packages may not be found!
Couldn't find index page for 'pytest' (maybe misspelled?)
Scanning index of all packages (this may take a while)
Reading https://pypi.org/simple/
Download error on https://pypi.org/simple/: [Errno -2] Name or service not known -- Some packages may not be found!
No local packages or working download links found for pytest
error: Could not find suitable distribution for Requirement.parse('pytest')
builder for '/nix/store/m4rnfxhryrd36ccsih9sqmd7hg1cn3f4-python3.6-pex-1.4.5.drv' failed with exit code 1
error: build of '/nix/store/24xkcyzzxnkj4d81404c5bqngw5dvlw0-python2.7-pex-1.4.5.drv', '/nix/store/m4rnfxhryrd36ccsih9sqmd7hg1cn3f4-python3.6-pex-1.4.5.drv' failed
```
</p></details>

Mic92 commented on pkgs/development/python-modules/pex/default.nix:
```diff
@@ -0,0 +1,28 @@
+  };
+
+  propagatedBuildInputs = [ setuptools wheel requests ];
+
```
Blind guess:

```nix
checkInputs = [ pytest ];
```

to fix the build issues:

mgeisler replied:
Thanks, I'll give it a try..

Mic92 commented:
@GrahamcOfBorg build pythonPackages.pex python3Packages.pex

GrahamcOfBorg commented:
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45497&attempt_id=be8b02cf-628f-4339-a8a3-7d32b3125e3d)

Attempted: pythonPackages.pex, python3Packages.pex

<details><summary>Partial log (click to expand)</summary><p>

```
warning: no previously-included files matching '*~' found under directory '*'
writing manifest file 'pex.egg-info/SOURCES.txt'
running build_ext

----------------------------------------------------------------------
Ran 0 tests in 0.000s

OK
/nix/store/7hy485pmia4h6mhfp1fhv4aimxfs72a4-python2.7-pex-1.4.5
/nix/store/0318k83l6lbpznfpxh4qlcmd1p6p9g3l-python3.6-pex-1.4.5
```
</p></details>

GrahamcOfBorg commented:
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45497&attempt_id=a7090f44-b9a2-4ed5-85c4-0803aa1aef79)

Attempted: pythonPackages.pex, python3Packages.pex

<details><summary>Partial log (click to expand)</summary><p>

```
Reading https://pypi.org/simple/twitter.common.dirutil/
Download error on https://pypi.org/simple/twitter.common.dirutil/: [Errno -2] Name or service not known -- Some packages may not be found!
Couldn't find index page for 'twitter.common.dirutil' (maybe misspelled?)
Scanning index of all packages (this may take a while)
Reading https://pypi.org/simple/
Download error on https://pypi.org/simple/: [Errno -2] Name or service not known -- Some packages may not be found!
No local packages or working download links found for twitter.common.dirutil<0.4.0,>=0.3.1
error: Could not find suitable distribution for Requirement.parse('twitter.common.dirutil<0.4.0,>=0.3.1')
builder for '/nix/store/1afi9zw1hlbn4mhz81k5b01ybf4291c7-python3.6-pex-1.4.5.drv' failed with exit code 1
error: build of '/nix/store/1afi9zw1hlbn4mhz81k5b01ybf4291c7-python3.6-pex-1.4.5.drv', '/nix/store/6kfijr4ma8gif468arl5v6p13xnw7w8p-python2.7-pex-1.4.5.drv' failed
```
</p></details>

GrahamcOfBorg commented:
Failure on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.45497&attempt_id=db0e0d6c-590e-4b33-961d-729cdc6d437a)

Attempted: pythonPackages.pex, python3Packages.pex

<details><summary>Partial log (click to expand)</summary><p>

```
Reading https://pypi.org/simple/twitter.common.dirutil/
Download error on https://pypi.org/simple/twitter.common.dirutil/: [Errno -2] Name or service not known -- Some packages may not be found!
Couldn't find index page for 'twitter.common.dirutil' (maybe misspelled?)
Scanning index of all packages (this may take a while)
Reading https://pypi.org/simple/
Download error on https://pypi.org/simple/: [Errno -2] Name or service not known -- Some packages may not be found!
No local packages or working download links found for twitter.common.dirutil<0.4.0,>=0.3.1
error: Could not find suitable distribution for Requirement.parse('twitter.common.dirutil<0.4.0,>=0.3.1')
builder for '/nix/store/7h3k3y90hywx49s8xdn89pkxqszhxgby-python3.6-pex-1.4.5.drv' failed with exit code 1
error: build of '/nix/store/7h3k3y90hywx49s8xdn89pkxqszhxgby-python3.6-pex-1.4.5.drv', '/nix/store/kw8hpc3pviqwlb0rwp68nw4sdw435qb5-python2.7-pex-1.4.5.drv' failed
```
</p></details>

Mic92 commented:
You can reproduce these problems locally, when enabling nix's sandbox.

mgeisler commented:
Thanks @Mic92 -- I'll have to experiment a bit more with this offline! You're completely right that I should get my local setup done so that I can find a way to reproduce what you see.

I'll ping you guys later when I have worked out the kinks... thanks for all the feedback until now!

mgeisler commented:
Hi guys -- I no longer work with Nix, so I won't be able to finish this. I hope the PR can be useful to someone else who wants to upgrade PEX.
